ribs/archives/_nearest_neighbor_archive.py

```diff
@@ -0,0 +1,489 @@
+                objective values to default to 0. However, if the user wishes to
+                associate an objective with each solution, this can be a
+                (batch_size,) array with objective function evaluations of the
+                solutions.
+            measures (array-like): (batch_size, :attr:`measure_dim`) array with
+                measure space coordinates of all the solutions.
+            fields (keyword arguments): Additional data for each solution. Each
+                argument should be an array with batch_size as the first
+                dimension.
+
+        Returns:
+            dict: Information describing the result of the add operation. The
+            dict contains the following keys:
+
+            - ``"status"`` (:class:`numpy.ndarray` of :class:`int`): An array of
+              integers that represent the "status" obtained when attempting to
+              insert each solution in the batch. Each item has the following
+              possible values:
+
+              - ``0``: The solution was not added to the archive.
+              - ``2``: The solution discovered a new cell in the archive.
+
+              Unlike in :class:`~ribs.archives.GridArchive`, there is no status
+              of 1 since solutions in this archive are never replaced.
+
+              To convert statuses to a more semantic format, cast all statuses
+              to :class:`AddStatus` e.g. with ``[AddStatus(s) for s in
+              add_info["status"]]``.
+
+            - ``"novelty"`` (:class:`numpy.ndarray` of :attr:`dtypes`
+              ["measures"]): The computed novelty of the solutions passed in. If
+              there were no solutions to compute novelty with respect to (e.g.,
+              the archive was empty), the novelty is set to infinity
+              (``numpy.inf``).
+
+        Raises:
+            ValueError: The array arguments do not match their specified shapes.
+            ValueError: ``objective`` or ``measures`` has non-finite values (inf
+                or NaN).
+        """
+        data = validate_batch(
+            self,
+            {
+                "solution":
+                    solution,
+                "objective":
+                    np.zeros(len(solution), dtype=self.dtypes["objective"])
+                    if objective is None else objective,
+                "measures":
+                    measures,
+                **fields,
+            },
+        )
+
+        reference_measures = self.data("measures")
+        if self._compare_to_batch:
+            reference_measures = np.concatenate(
+                (reference_measures, data["measures"]), axis=0)
+
+        if self.compare_to_batch:
+            # In this case, solutions from the batch will be included in the
+            # nearest neighbors. Thus, the first neighbor of each solution will
+            # be the solution itself, so we need to compute an extra neighbor
+            # and exclude the first neighbor.
```

Comment:
I'm not sure the statement "the first neighbor of each solution will be the solution itself".

A counterexample:
In the unlikely event that two solutions share the exact same point measure space and unless `cKDTree` behaves differently than other sort methods, the lower-indexed one will appear first. This means the "zeroth" neighbor is not always the solution itself.

In an even more unlikely scenario where `k+1` other solutions occupy the same point in measure space *and* all have lower indices than the solution itself, then the solution won't even appear in the k-nearest neighbors.

---

I went about handling this by filtering the return of `cKDTree.query` with something like `knn_indices != (len(self) + np.arange(batch_size))[:, None]`

Reply:
If two solutions are at the same point, then one of them should indeed be marked as the first neighbor and will be excluded, and the other should be marked as a neighbor that should be included.

Anyway, I think this `compare_to_batch` behavior is not a good idea because it essentially forces the batch to be spread out. I will remove it.